File: doc/clients/csharp/updatecheck.cs
// Usage:
// var client = new UpdateCheckClient("https://license.example.com");
// var updateInfo = await client.CheckForUpdateAsync(licenseKey, machineCode, "1.0.0");
// if (updateInfo.UpdateAvailable)
//     Console.WriteLine($"Update available: {updateInfo.LatestVersion} - {updateInfo.DownloadURL}");

using System.Net.Http.Json;

public record LicenseInfoRequest(string MachineCode, string InstalledVersion);

public record LicenseInfoResponse(
    string CustomerName,
    string ProductGUID,
    string ProductName,
    int LicenseCount,
    int LicensesAvailable,
    string ExpirationDate,
    string MaintExpirationDate,
    string MaxProductVersion,
    string LatestVersion,
    Dictionary<string, object>? Features
);

public record ProductVersionResponse(
    string ProductGUID,
    string LatestVersion,
    string DownloadURL
);

public record UpdateInfo(
    bool UpdateAvailable,
    string CurrentVersion,
    string LatestVersion,
    string? DownloadURL
);

public class UpdateCheckClient
{
    private readonly HttpClient _client;
    private readonly string _baseUrl;

    public UpdateCheckClient(string baseUrl)
    {
        _baseUrl = baseUrl;
        _client = new HttpClient();
    }

    /// <summary>
    /// Reports the installed version to the server and returns license info.
    /// </summary>
    public async Task<LicenseInfoResponse> UpdateInstalledVersionAsync(
        string licenseKey,
        string machineCode,
        string installedVersion)
    {
        var request = new LicenseInfoRequest(machineCode, installedVersion);

        var response = await _client.PutAsJsonAsync($"{_baseUrl}/api/v1/license/{licenseKey}", request);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadFromJsonAsync<LicenseInfoResponse>()
            ?? throw new InvalidOperationException("Empty response");
    }

    /// <summary>
    /// Retrieves the latest version and download URL for a product.
    /// </summary>
    public async Task<ProductVersionResponse> GetProductVersionAsync(string productGUID)
    {
        var response = await _client.GetAsync($"{_baseUrl}/api/v1/productver/{productGUID}");
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadFromJsonAsync<ProductVersionResponse>()
            ?? throw new InvalidOperationException("Empty response");
    }

    /// <summary>
    /// Reports the installed version and checks if an update is available.
    /// Returns update info including the download URL if an update is available.
    /// </summary>
    public async Task<UpdateInfo> CheckForUpdateAsync(
        string licenseKey,
        string machineCode,
        string installedVersion)
    {
        // Step 1: Report installed version and get license info
        var licenseInfo = await UpdateInstalledVersionAsync(licenseKey, machineCode, installedVersion);

        // Step 2: Check if update is available
        if (LicenseClient.CompareVersions(installedVersion, licenseInfo.LatestVersion) >= 0)
        {
            return new UpdateInfo(
                UpdateAvailable: false,
                CurrentVersion: installedVersion,
                LatestVersion: licenseInfo.LatestVersion,
                DownloadURL: null
            );
        }

        // Step 3: Get download URL
        var productInfo = await GetProductVersionAsync(licenseInfo.ProductGUID);

        return new UpdateInfo(
            UpdateAvailable: true,
            CurrentVersion: installedVersion,
            LatestVersion: licenseInfo.LatestVersion,
            DownloadURL: productInfo.DownloadURL
        );
    }
}

File: doc/clients/go/updatecheck.go
// sample implementation, do not build or test
//go:build ignore

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
)

type LicenseInfoRequest struct {
	MachineCode      string `json:"machineCode"`
	InstalledVersion string `json:"installedVersion"`
}

type LicenseInfoResponse struct {
	CustomerName        string         `json:"CustomerName"`
	ProductGUID         string         `json:"ProductGUID"`
	ProductName         string         `json:"ProductName"`
	LicenseCount        int            `json:"LicenseCount"`
	LicensesAvailable   int            `json:"LicensesAvailable"`
	ExpirationDate      string         `json:"ExpirationDate"`
	MaintExpirationDate string         `json:"MaintExpirationDate"`
	MaxProductVersion   string         `json:"MaxProductVersion"`
	LatestVersion       string         `json:"LatestVersion"`
	Features            map[string]any `json:"Features"`
}

type ProductVersionResponse struct {
	ProductGUID   string `json:"ProductGUID"`
	LatestVersion string `json:"LatestVersion"`
	DownloadURL   string `json:"DownloadURL"`
}

type UpdateInfo struct {
	UpdateAvailable  bool
	CurrentVersion   string
	LatestVersion    string
	DownloadURL      string
}

// UpdateInstalledVersion reports the installed version to the server and returns license info.
func UpdateInstalledVersion(baseURL, licenseKey, machineCode, installedVersion string) (*LicenseInfoResponse, error) {
	reqBody := LicenseInfoRequest{
		MachineCode:      machineCode,
		InstalledVersion: installedVersion,
	}

	body, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}

	req, err := http.NewRequest("PUT", baseURL+"/api/v1/license/"+licenseKey, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("update failed: %s", resp.Status)
	}

	var result LicenseInfoResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decode response: %w", err)
	}

	return &result, nil
}

// GetProductVersion retrieves the latest version and download URL for a product.
func GetProductVersion(baseURL, productGUID string) (*ProductVersionResponse, error) {
	resp, err := http.Get(baseURL + "/api/v1/productver/" + productGUID)
	if err != nil {
		return nil, fmt.Errorf("send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("get product version failed: %s", resp.Status)
	}

	var result ProductVersionResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decode response: %w", err)
	}

	return &result, nil
}

// CheckForUpdate reports the installed version and checks if an update is available.
// Returns update info including the download URL if an update is available.
func CheckForUpdate(baseURL, licenseKey, machineCode, installedVersion string) (*UpdateInfo, error) {
	// Step 1: Report installed version and get license info
	licenseInfo, err := UpdateInstalledVersion(baseURL, licenseKey, machineCode, installedVersion)
	if err != nil {
		return nil, fmt.Errorf("update installed version: %w", err)
	}

	info := &UpdateInfo{
		CurrentVersion: installedVersion,
		LatestVersion:  licenseInfo.LatestVersion,
	}

	// Step 2: Check if update is available
	if CompareVersions(installedVersion, licenseInfo.LatestVersion) >= 0 {
		info.UpdateAvailable = false
		return info, nil
	}

	// Step 3: Get download URL
	productInfo, err := GetProductVersion(baseURL, licenseInfo.ProductGUID)
	if err != nil {
		return nil, fmt.Errorf("get product version: %w", err)
	}

	info.UpdateAvailable = true
	info.DownloadURL = productInfo.DownloadURL

	return info, nil
}
